web_client/actions.jsx

```diff
@@ -223,6 +223,101 @@ export const enableAnalysisPage = (page) => promiseAction(
   }
 );
 
+export const ensurePrivateDirectory = (() => {
+  let singleton;
+
+  return (args) => {
+    if (!singleton) {
+      singleton = ensureUserDirectory({
+        name: 'Private',
+        description: 'Private Directory',
+        ...args
+      });
+    }
+
+    return singleton;
+  };
+})();
+
+export const ensureScratchDirectory = (() => {
+  let singleton;
+
+  return (args) => {
+    if (!singleton) {
+      singleton = ensureUserDirectory({
+        name: '__osumo_tmp',
+        description: 'Temporary Staging Space',
+        ...args
+      });
+    }
+
+    return singleton;
+  };
+})();
+
+export const ensureUserDirectory = (args) => {
+  let {
+    description,
+    name,
+    returnModel,
+    user
+  } = args;
+
+  return promiseAction(
+    (dispatch, getState) => {
```

Comment:
tl; dr Pretty soon, none of what I wrote below is going to matter.

So, consider these two points about an "idealized" implementation of this function:

1 - Multiple concurrent calls for the same directory are not erroneous; the point is to ensure that the *same* promise chain is returned to each caller in such a case.  This approach avoids race conditions in a way that doesn't require that every call site be aware of each other.  In fact, the same promise chain would be ideally returned even long after having been resolved (effectively making subsequent calls a noop that immediately returns a promise that resolves immediately).  It would be best if you didn't have to go over the wire to ensure the presence of a directory that you've already checked for in a previous call.

2 - A new promise chain is created for every distinct directory queried -- on the first such call for that directory.  This idealized implementation is basically a memoized function returning a promise chain that is unique to the input directory.

The code being discussed differs from the ideal in that I specifically chose not to memoize it, because doing so would require managing a table of promise chains that would grow, possibly without bound, and in a way very unlikely to be obvious to other developers.  I made exceptions for the user's Private folder and the scratch space folder, since OSUMO specifically needs those two folders, and figured that leaving other call sites for other directories to coordinate among themselves was a pragmatic alternative to caching or full memoization.

In any case, @chaitanya2334 and I actually figured out what was the issue with the R scripts, today, and my plan for addressing that issue will mostly supplant this part of the code.